runtime/common/src/parachains.rs

```diff
@@ -99,13 +112,65 @@ impl<AccountId, T: Currency<AccountId>> ParachainCurrency<AccountId> for T where
 /// Interface to the persistent (stash) identities of the current validators.
 pub struct ValidatorIdentities<T>(rstd::marker::PhantomData<T>);
 
+#[derive(RuntimeDebug, Encode, Decode)]
+#[derive(Clone, Eq, PartialEq)]
+pub struct DoubleVoteReport {
+	/// Identity of the double-voter.
+	pub identity: ValidatorId,
+	/// First vote of the double-vote.
+	pub first: (Statement, ValidityAttestation),
+	/// Second vote of the double-vote.
+	pub second: (Statement, ValidityAttestation),
+}
+
+impl DoubleVoteReport {
+	fn verify<H: AsRef<[u8]>>(
+		&self,
+		parent_hash: H,
+	) -> Result<(), ()> {
+		let first = self.first.clone();
+		let second = self.second.clone();
+
+		Self::verify_vote(first, &parent_hash, self.identity.clone())?;
+		Self::verify_vote(second, &parent_hash, self.identity.clone())?;
+
+		Ok(())
+	}
+
+	fn verify_vote<H: AsRef<[u8]>>(
+		vote: (Statement, ValidityAttestation),
+		parent_hash: H,
+		authority: ValidatorId,
+	) -> Result<(), ()> {
+		let (statement, attestation) = vote;
+
+		let (payload, sig) = match attestation {
+			ValidityAttestation::Implicit(sig) => {
+				let payload = localized_payload(statement, parent_hash);
```

Comment:
i.e. a `ValidityAttestation::Implicit` corresponds directly to a `Statement::Candidate`. and a `ValidityAttestation::Explicit` corresponds to a `Statement::Valid`. However, the `statement` isn't checked against that type, so the code doesn't make sense. In this case, a `Statement, Signature` pair would make most sense.